### ISSUE TITLE
Keeping its Tatoeba data up to date requires a lot of unnecessary downloading

### ISSUE THREAD

**Issue by LBeaudoux** [Tatoeba/tatoeba2#2680]
## Problem

Some developers want to enjoy the most recent data from Tatoeba on their devices. This is why they download the export files they need on a weekly basis. 

Each time, they have to download the whole dataset once again, while in most cases, only a small proportion of the data is added, modified or removed between two updates. As @mramosch already [pointed out](https://github.com/Tatoeba/tatoeba2/issues/2648#issuecomment-798768763), this unnecessary downloading is particularly problematic in mobile environments that are subject to data quotas.

## Proposed solution

@jiru  has already highlighted [a solution](https://github.com/Tatoeba/tatoeba2/issues/2648#issuecomment-798898766) to this problem: for each file exported, another file containing only the differences to the previous version is generated.

The easiest way to build these "diff files" may consist in using a command-line tool that compare similar CSV files to each other. However, such an operation applied on large files easily consumes several GB of memory.  And this issue would get worse with the growth of the Tatoeba corpus.

Performing the comparison directly in MariaDB would surely be more efficient in terms of memory. Every week we could copy the results of the export queries into newly created tables in the database. Two versions of the same export table could then be compared thanks to joins on primary keys, and the differences found saved as a CSV or JSON file. For a week, the latest export tables could be conveniently used to generate [custom parallel corpora](https://github.com/Tatoeba/tatoeba2/pull/2636). Finally, after being compared to a newer version of themselves, they would finally become useless and consequently deleted.

**Comment by mramosch** [comment]
The only ‘proper’ way to stay up-to-date is an API.

I for example have added an ‘Extended native speaker quality evaluation system’ for sentences to my App that needs the most recent data in order to not waste man power only because native contributors are evaluating sentences that already got enough evaluation input in order to appear as 100% ‘curated’, but simply don’t show up us such because the update comes only once a week.

By that time, because of this possible ‘over-evaluation’ of some sentences, an enormous amount of contributions would have been submitted in vain that could instead have been used to bring other - not yet sufficiently evaluated sentences - to a higher curation level.

Living on huge file downloads that have to be checked, sanitized, organized, pushed over and stored on every user’s device in their entirety is a real bad choice for that use case. Only real-time access to this data could solve the disk-space, mobile data quota and not being up-to-date problem with all its consequences...

**Comment by LBeaudoux** [comment]
It seems to me that we are dealing with two types of scenarios here.

### 'Hot differences' scenario

The third party application needs to interact tightly and almost in real time with Tatoeba. The communication is conducted  through a REST API. To update itself, the application queries the sentences, links, tags, users, audios, etc. that have been added, modified or removed since a datetime. 

For these requests to work properly, every change made on a database record would have to be timestamped somewhere (and easily accessible). It doesn't seem to me that this is the case right now. Besides, in order to keep the responses' sizes below a few megabytes, the period of time between updates should not be too long: one week may be a suitable maximum.

### 'Cold differences' scenario

The third party application doesn't need the latest contributions made in Tatoeba but still wants to take advantage of the improvements of the corpus. For this, it automatically triggers updates once a week (when it is online). The first time, the complete export files are downloaded. The following times, only the packages containing the incremental differences of the missed weeks are downloaded. 

To make this possible on the server side, a copy of the export views could be saved each week into a dedicated database. Note that if the changes in the tables have not all been timestamped, it is still possible to deduce them by comparing the tables from one week to the next. 

The 'export database' would then be optimized to generate:
- all complete monolingual and multilingual exports (CSV files)
- all incremental differences in the monolingual and multilingual exports (JSON files)
- custom exports on demand

**Comment by mramosch** [comment]
Regarding the 'Hot differences' scenario

Like with new sentence contributions the server updates itself within 10-15 minutes.

If changes to a user profile (language skills etc.) follow the same procedure there is no need of a timestamp if you don’t have to maintain an inApp duplicate of a server-data-subset.

In my example from above I just ask an endpoint for the user of every received translation after a request/response to a sentence page.

Correlating this user info with the skill level field in a 300kB UserDetails file from the download page is much more reasonable and maintainable than having to deal with 900 GB...

Of course this would be even easier if the JSON dictionary in the sentence-page response HTML (that is used to init the AngularApp) already carried that user information for every translation, like it already does for the source sentence.

This would also make an additional network call unnecessary, no synchronization issues between UI and asynchronous availability of data chunks.

But this would be an addition useful to my application, so if there are no other developers needing the same extension of the JSON blob, I don’t see a chance to get this anytime soon...

**Comment by LBeaudoux** [comment]
@mramosch, thanks for your brief actual comment on this issue:

> Like with new sentence contributions the server updates itself within 10-15 minutes.
> 
> If changes to a user profile (language skills etc.) follow the same procedure there is no need of a timestamp if you don’t have to maintain an inApp duplicate of a server-data-subset.

Unfortunately, I'm afraid I didn't understand your explanation very well... Could you please try to explain more clearly why you think that the parsing of addition, modification and deletion timestamps is useless to identify the latest changes?

**Comment by mramosch** [comment]
My point simply was that if I access every item in real time from an endpoint on the server and I don’t have to maintain a copy of the whole DB on the client, then there is no need for diffing out or synchronizing data-sets on different devices.

If I contribute a new sentence it will be indexed and available in the search engine in about 10 minutes.

So if I seek for truth, truth is on the server... (at least after 10 minutes)

Maybe parameter changes could be available immediately because the records already exist and there is no new indexing necessary. So unintended redundant work on the same item is pretty unlikely.

Just like two or more (n) people evaluating a sentence that only needs one evaluation. Unless they all happen to request the same sentence at the same time and submit their evaluation at the same time. Then some man power would have been wasted because 2 or more (n) sentences could have received an evaluation instead of the redundant multi evaluation of this first sentence that only needed one single evaluation.

**Comment by LBeaudoux** [comment]
> My point simply was that if I access every item in real time from an endpoint on the server when needed and I don’t have to maintain a copy of the whole DB on the client, then there is no need for diffing out or synchronizing data-sets on different devices.

@mramosch I'm afraid we're talking about different issues here.  I started this thread to brainstorm about a way to provide lists of _**all**_ recent changes made on the corpus. I have the feeling that you are only interested in whether a **_particular_** object (i.e. already identified) has undergone changes. 

This is another topic that I think would be better addressed in the [broader issue about the API](https://github.com/Tatoeba/tatoeba2/issues/2669).

**Comment by mramosch** [comment]
Sorry, but then I misunderstood your intention in the first place...

But your initial argument linked twice to my thread where I discussed with @jiru the necessity of either an extension of the JSON object in the HTML response or a decent API in order to avoid downloading, maintaining and synchronizing data at all.

And that having to download files that are already outdated from the moment they were created is only a poor replacement for realtime access - and that some kind of diff files would be a more or less ‘acceptable’ interim solution until we get realtime access to the server. But I wouldn’t consider jiru’s suggestion in that other thread as a ‘solution’ to the problem at hand there.


So with all your references to that thread I thought you are jumping on the same band wagon and wasn’t aware that you in reality are embracing the idea of maintaining an inApp copy of the DB and diff files as a cure to your synchronization and band width issues. Well, in your case jiru’s proposal could help out...


And speaking of brainstorming: what would your needs as a ‘developer’ be? Without talking about your idea of any solution, what language problem is your Application solving, and which data does it need and where, why and when?

As I understand it, this is jiru’s preferred approach where HE, with all his knowledge about the backend and the inner workings of Tatoeba in general, tries to approach the problem in his own pace and later on asks for further input and ideas for solutions.

And I would be interested, too ;-)

The more app developers need and ask for better technologies to access the corpus the sooner we’ll get the tools we need.

Curious about your upcoming answer!

Take care...

**Comment by LBeaudoux** [comment]
> The more app developers need and ask for better technologies to access the corpus the sooner we’ll get the tools we need.

@mramosch I'm sorry to tell you this, but it seems to me that things don't work exactly as you think they are. Actually, the best way to improve Tatoeba's software is to  submit pull requests. The issues mainly help the developer who contributes to better understand the technical challenges to overcome. By not reading the posts carefully enough, by answering hastily with long off-topic comments, by criticizing an experienced and dedicated maintainer, I doubt that you have helped your cause in any way...

I would advise you to take a look at the other issues, if you want to get a better idea of the guidelines that this community follows.

I am sorry because it's probably not the kind of answer you were expecting, but sometimes things have to be said a bit bluntly to be heard. I sincerely hope that our next conversation can be more constructive.

**Comment by mramosch** [comment]
> Some **_developers_** want to enjoy the most recent data from Tatoeba on their devices...
> As mramosch already pointed out, this unnecessary downloading... 

———

I didn’t intend to insult or criticize you at all. If it came across this way I apologize.

I was reading your posts very attentively but I just misunderstood the direction you were pitching the thread to because you were referencing me in the presentation of the problem and the problem I see as a ‘**_developer_**’ is having to deal with offline downloads at all.

But I still don’t see your problem as a ‘**_developer_**’ - 

 • the “**what**” is your application trying to achieve?
 • the “**why**” you would still prefer (external) downloads if you could have direct inApp access to data-points?

That’s what I would appreciate to know - if you can still bear with me and spare some time.

**Comment by LBeaudoux** [comment]
> I didn’t intend to insult or criticize you at all. 

I am not a maintainer of Tatoeba. In my previous post, I was referring to your criticism of jiru.

> “what” is your application trying to achieve?

I am not developing any particular application. I'm just trying to contribute to the enhancement of exports that jiru has undertaken lately. 

>  “why” you would still prefer (external) downloads if you could have direct inApp access to data-points?

Sometimes, a whole corpus need to be parsed. The examples are manifold: identifying the best sentences, training machine translation models, infering statisticals insights, etc.

**Comment by mramosch** [comment]
I totally agree that for some tasks downloading bulk data is a necessity. We can see their merit e.g. in the wonderful Statistics-Link we get every week on the Wall...

But neither did I see the word ‘enhancement’ in your text nor as a label for the thread, so I was totally misled by the choice of words in your introduction in the initial post.

• developer - devices - mobile environment

And as you were citing me on mobile environments

> ...that are subject to data quotas.

I mistook you as an ‘Application Developer for Mobile Devices’ ;-) and for that matter your sentence

> @jiru has already highlighted a solution to this problem

didn’t resonate with me at all, because I don’t see bulk downloading as a permanent solution for most of  mobile development. All this thread was driving at was about finding alternative solutions to bulk-downloading, which can clearly be seen in https://github.com/Tatoeba/tatoeba2/issues/2648#issuecomment-798912594 .

• establishment of an API
• extension of the JASON object in the HTML response


But my bad, obviously...


———


And as to my alleged criticism of @jiru :

I am not aware that I did such thing. I have the highest respect for him and he has helped me out several times now with his clear analytic dissection of some problems I encountered.

So in case I unwittingly did I apologize again (to @jiru of course ;-)

**Comment by LBeaudoux** [comment]
@mramosch I'm glad we finally understand each other better. 

I now suggest that we hide all off-topic content from this thread to avoid boring other readers with this argument. I let you start, then I'll rewrite my first post and try to make it clearer.

**Comment by mramosch** [comment]
Hide? Delete? Difference...?

**Comment by LBeaudoux** [comment]
Hiding comments looks fine. More details [here](https://docs.github.com/en/communities/moderating-comments-and-conversations/managing-disruptive-comments)

**Comment by mramosch** [comment]
Done. Even my first comment will appear to be odd in case you clarify your initial post, so I hid it as well.

If it still appears applicable after your edits I will unhide it again. The rest beneath that - is history...

**Comment by LBeaudoux** [comment]
@mramosch I'm done with the first post, I let you modify the second one. I would like to keep the third post as it is. But I now think that we should just delete all following posts altogether.

**Comment by mramosch** [comment]
I also like the 3rd post because it goes even one level deeper, however my preceding 2nd comment doesn’t contribute anything to bridge between the 1st and 3rd one.

So if it’s ok with you that I also remove the 2nd comment and your 3rd one seems to be a follow up to your own 1st one...

Maybe you want to integrate the 3rd one in your original post or do you rather want a proper bridge in my 2nd comment and keep your 1st and 3rd one separated?

**Comment by LBeaudoux** [comment]
I think your reminder of the need for an API to achieve (quasi) real time is a good bridge between my two posts.

**Comment by mramosch** [comment]
But it still sounds off-topic because it has nothing to do or does in no way continue or pick up the thread of your initial proposal. It just sounds like some random contradiction although I know that for some use cases bulk downloads are still useable...

There should be a least an initial bridging sentence that makes it plausible to steer into the direction of giving yet another reminder for the importance of an API.

**Comment by LBeaudoux** [comment]
Feel free to delete your second post if you think it does not bring enough value. I can change the intro of the third post to create a smoother transition...

**Comment by mramosch** [comment]
Have a look at the 2nd post now, please.

Does it seem appropriate to hijack your thread in that fashion or do you prefer a removal?

**Comment by LBeaudoux** [comment]
I like your second post. Keep it as it is. I think we can finally delete the hidden posts now...

**Comment by mramosch** [comment]
No worries, done.

**Comment by jiru** [comment]
While I do think we need to solve this issue, I’d rather start implementing a solution that answers an actual real-world use case instead of just thinking about "some third-party application" that "might need" this or that. Otherwise I am afraid we are going to implement something that works but will be little use in practice. In other words, start simple and don’t think too big.

@LBeaudoux If you are creating a third-party application, please share your use case with us. For example, after you downloaded the CSV files, do you convert the data into a different format?

Such issue has certainly been tackled by other projects. The first that comes to my mind is the OSMAnd app. That app downloads large chunks of OpenStreetMap data to display offline maps of the world. You can basically update the maps manually every now and then by downloading the new stuff (similar to your 'Cold differences' scenario). There is also a feature called "OSMAnd Live" that keeps your local database constantly up to date by subscribing to a feed of updates (similar to your 'Hot differences' scenario).

For this to work, I have the feeling that we’d need to change the way we store our data under the hood, to make it more "transactional", similar to mysql’s binlog or git revisions. Every single change to the corpus should have a "revision number", and you should be able to go to a newer revision by applying the log of changes to your own copy of the corpus, from an arbitrary revision number. While we do have a log of changes (the `contributions` table), not only it does not record all the changes, but it contains a lot of inconsistencies because of crashes or manual edits of the mysql database. Even without bugs and crashes, relying on mysql makes the whole system error-prone because we can break it just by running some SQL queries. Instead, we’d need a database system with a revision feature built-in.

Related: #1633.

**Comment by mramosch** [comment]
—

> @jiru: While I do think we need to solve this issue, I’d rather start implementing a solution that answers an actual real-world use case instead of just thinking about "some third-party application" that "might need" this or that. Otherwise I am afraid we are going to implement something that works but will be little use in practice. In other words, start simple and don’t think too big.

—

I’ll post my comment in a new thread in order not to hijack this one here...

**Comment by LBeaudoux** [comment]
> If you are creating a third-party application, please share your use case with us. 

I don't have any particular project in progress at the moment, but I will try to illustrate some real (or at least plausible) examples.

**A  flashcard app could enjoy incremental data releases**

Let's say that as a developer, I have  built a mobile flashcard app. When a user wants to review a word, he adds it to his vocabulary list and flashcards of good Tatoeba sentence pairs containing this word are generated automatically. The app works offline thanks to a local SQLite database storing all sentence pairs above a minimum quality score. 
As of today,  a data update would consist in (re)downloading all sentences and links related to a language pair. Then, all sentence pairs would have to be (re)scored to discard low-quality pairs. By comparing new export files with their former versions to identify differences, the process can be optimized at the cost of a larger peak memory footprint. But ideally, only necessary data should be downloaded, processed and stored on the user device. 
This is why providing incremental data releases would be a significant improvement.

**Tatominer could mirror Tatoeba in real-time thanks to a transaction feed**

As a developer, I launched [Tatominer](https://tatominer.netlify.app/) to highlight the vocabulary searched on Tatoeba that is poorly covered. 
For now, when a user adds a sentence on Tatoeba containing a word listed by Tatominer, his contribution is not immediately taken into account by Tatominer. One contributor [noted](https://tatoeba.org/eng/wall/show_message/35535#!#message_35535) that immediately seeing the impact of his contributions would help him keep track of where he is, and also give him a feeling of accomplishment.
Providing a feed of the transactions made on the Tatoeba database minute by minute would enable Tatominer to track these new contributions more closely.

> That app downloads large chunks of OpenStreetMap data to display offline maps of the world.

OpenStreetMap daily, hourly and even minutely [publishes](https://planet.osm.org/) diff files. Tools like [Osmium](https://osmcode.org/osmium-tool/) have been developed to process these diff files. This could indeed be a source of inspiration for us.

>  I have the feeling that we’d need to change the way we store our data under the hood, to make it more "transactional", similar to mysql’s binlog

I stumbled upon a [daemon](https://maxwells-daemon.io/) that reads MySQL binlogs and writes row updates as JSON that can be later used to replicate the source database. It seems to me that this tool could help us to pipe a stream of transaction events to diff files. This transaction stream could also serve as a base for new Tatoeba features (e.g. custom news feeds).

**Comment by jiru** [comment]
About the flashcard app, I can imagine the app would initially download CSV files, insert the sentences into its database and remove the CSV files (to save disk space). Then, to update its database, the app would need to get a list of new insertions, removals and updates of sentence texts, a list of insertions and removals of links.

About Tatominer, maybe it could benefit from Manticore’s [percolate queries](https://manual.manticoresearch.com/Searching/Percolate_query) (we’d have to switch from plain indexes to real-time indexes though). If we somehow serve percolate query results through an API, Tatominer could receive all new sentences that match the watched keywords only.

**Comment by jiru** [comment]
Maybe we could use the [data versioning feature of mariadb](https://mariadb.com/fr/resources/blog/automatic-data-versioning-in-mariadb-server-10-3/).

**Comment by LBeaudoux** [comment]
> Maybe we could use the system-versioned tables feature of mariadb.

It seems possible to compare two weekly versions of a system-versioned table within MariaDB.  But  exporting diff files with this method would be heavy, especially for shorter time periods  (e.g. hourly or minutely diffs).

The more I think about it, the more I am convinced that we should be inspired by the OpenStreetMap example and periodically publish the Tatoeba transaction history as JSON files. This approach would cover more use cases and scale better than the other [CDC](https://en.wikipedia.org/wiki/Change_data_capture) options we have discussed so far. 

Obviously, this would be an ambitious goal that would need to be broken down into several milestones. A first step could be to generate a daily transaction file to experiment with [Maxwell's Daemon](https://maxwells-daemon.io/).

This piece of software can read  binlogs on the fly and write row updates as JSON (no migration required) :
```sh
bin/maxwell --user='maxwell' --password='XXXXXX' --host='127.0.0.1' --producer=stdout > today_transactions.json
```
For example, when a new translation is added:
```mysql
  mysql> INSERT INTO `tatoeba`.`sentences_translations` SET sentence_id = 269101, translation_id = 9965114;
  maxwell: {
    "database": "tatoeba",
    "table": "sentences_translations",
    "type": "insert",
    "ts": 1449786310,
    "xid": 940752,
    "commit": true,
    "data": { "sentence_id": 269101, "translation_id": 9965114}
  }

```

**Comment by jiru** [comment]
I second your suggestion.

By the way, the weekly CSV files are mostly consistent to one another but not exactly. For example, if a new translation is added while sentences.csv is being generated, the sentence won’t show up because it was added after the generation started, but I believe it will show up in links.csv because that file is generated later on. This happens because the files are not generated in a single transaction. It’s probably easy to fix with a BEGIN/COMMIT though.
